src/build_system.ml

```diff
@@ -1583,15 +1583,47 @@ let on_load_dir ~dir ~f =
       add_build_dir_to_keep t ~dir;
     p.lazy_generators <- f :: lazy_generators
 
-let eval_pred ~dir pred =
-  let t = t () in
-  Path.Set.fold (targets_of t ~dir) ~init:[] ~f:(fun path acc ->
-    let fn = Path.basename path in
-    if Predicate.test pred fn then
-      fn :: acc
-    else
-      acc)
-  |> List.rev
+let eval_pred_def =
+  let module Input = struct
+    type t = Path.t * string Predicate.t
+
+    let to_dyn (path, pred) =
+      Dyn.Tuple [Path.to_dyn path; Predicate.to_dyn pred]
+
+    let to_sexp t = Dyn.to_sexp (to_dyn t)
+    let equal = Tuple.T2.equal Path.equal Predicate.equal
+    let hash = Tuple.T2.hash Path.hash Predicate.hash
+  end
+  in
+  let module Output = struct
+    type t = string list
+    let to_sexp = let open Sexp.Encoder in list string
+    let equal = List.equal String.equal
+  end
+  in
+  Memo.create "eval-pred"
+    ~doc:"Evaluate a predicate in a directory"
+    ~input:(module Input)
+    ~output:(Allow_cutoff (module Output))
+    ~visibility:Hidden
+    Sync
+    None
```

Comment:
I noticed that you often use `set_impl`. Any particular reason for doing so? It is slightly less efficient as it adds an indirection to the function call

Reply:
Only because it looks aesthetically better to me. Separating the definition from the implementation. I wasn't aware it's less efficient. I will do it in 1 step.

Reply:
Ok. Maybe we could keep the aesthetic element by partially applying the function:

```ocaml
let eval_pred_def = ... Memo.create ... Sync

let eval_pred_def =
  let f ... 
  in
  eval_pred_def (Some f)